diff --git a/quickjs.c b/quickjs.c
index fa5a7a9..2a91508 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -3846,6 +3846,97 @@ JSClassID JS_GetClassID(JSValue v)
     return p->class_id;
 }
 
+JS_BOOL JS_IsArrayBuffer(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_ARRAY_BUFFER;
+}
+
+JS_BOOL JS_IsDataView(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_DATAVIEW;
+}
+
+JS_BOOL JS_IsDate(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_DATE;
+}
+
+JS_BOOL JS_IsRegExp(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_REGEXP;
+}
+
+JS_BOOL JS_IsPromise(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_PROMISE;
+}
+
+JS_BOOL JS_IsProxy(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_PROXY;
+}
+
+JS_BOOL JS_IsMap(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_MAP;
+}
+
+JS_BOOL JS_IsSet(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_SET;
+}
+
+JS_BOOL JS_IsWeakMap(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_WEAKMAP;
+}
+
+JS_BOOL JS_IsWeakSet(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_WEAKSET;
+}
+
+JS_BOOL JS_IsWeakRef(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_WEAK_REF;
+}
+
+JS_BOOL JS_IsAsyncFunction(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_ASYNC_FUNCTION;
+}
+
+JS_BOOL JS_IsGeneratorFunction(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_GENERATOR_FUNCTION;
+}
+
+JS_BOOL JS_IsGenerator(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_GENERATOR;
+}
+
+JS_BOOL JS_IsArguments(JSValueConst v)
+{
+    JSClassID class_id = JS_GetClassID(v);
+    return class_id == JS_CLASS_ARGUMENTS || class_id == JS_CLASS_MAPPED_ARGUMENTS;
+}
+
+JS_BOOL JS_IsMapIterator(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_MAP_ITERATOR;
+}
+
+JS_BOOL JS_IsSetIterator(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_SET_ITERATOR;
+}
+
+JS_BOOL JS_IsModuleNamespace(JSValueConst v)
+{
+    return JS_GetClassID(v) == JS_CLASS_MODULE_NS;
+}
+
 BOOL JS_IsRegisteredClass(JSRuntime *rt, JSClassID class_id)
 {
     return (class_id < rt->class_count &&
diff --git a/quickjs.h b/quickjs.h
index 476d735..d964b49 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -770,6 +770,24 @@ JS_BOOL JS_SetConstructorBit(JSContext *ctx, JSValueConst func_obj, JS_BOOL val)
 
 JSValue JS_NewArray(JSContext *ctx);
 int JS_IsArray(JSContext *ctx, JSValueConst val);
+JS_BOOL JS_IsArrayBuffer(JSValueConst val);
+JS_BOOL JS_IsDataView(JSValueConst val);
+JS_BOOL JS_IsDate(JSValueConst val);
+JS_BOOL JS_IsRegExp(JSValueConst val);
+JS_BOOL JS_IsPromise(JSValueConst val);
+JS_BOOL JS_IsProxy(JSValueConst val);
+JS_BOOL JS_IsMap(JSValueConst val);
+JS_BOOL JS_IsSet(JSValueConst val);
+JS_BOOL JS_IsWeakMap(JSValueConst val);
+JS_BOOL JS_IsWeakSet(JSValueConst val);
+JS_BOOL JS_IsWeakRef(JSValueConst val);
+JS_BOOL JS_IsAsyncFunction(JSValueConst val);
+JS_BOOL JS_IsGeneratorFunction(JSValueConst val);
+JS_BOOL JS_IsGenerator(JSValueConst val);
+JS_BOOL JS_IsArguments(JSValueConst val);
+JS_BOOL JS_IsMapIterator(JSValueConst val);
+JS_BOOL JS_IsSetIterator(JSValueConst val);
+JS_BOOL JS_IsModuleNamespace(JSValueConst val);
 
 JSValue JS_NewDate(JSContext *ctx, double epoch_ms);
 
