diff --git a/quickjs.c b/quickjs.c
index fa5a7a9..94433de 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -622,7 +622,7 @@ typedef enum {
 } JSClosureTypeEnum;
 
 typedef struct JSClosureVar {
-    JSClosureTypeEnum closure_type : 3;
+    uint8_t closure_type : 3; /* see JSClosureTypeEnum */
     uint8_t is_lexical : 1; /* lexical variable */
     uint8_t is_const : 1; /* const variable (is_lexical = 1 if is_const = 1 */
     uint8_t var_kind : 4; /* see JSVarKindEnum */
